### DISCUSSION TITLE
useFieldArray, YUP and Material UI Problems

### DISCUSSION THREAD

**Discussion by DrBlackBird** [react-hook-form/react-hook-form#5447] [Q&A] [answered]
Hi everybody!

I´m pretty new to react-hook-form but I like it so far. I´ve managed some scenarios with simple validation together with material-ui and its working pretty well. 

Right now I´m having a problem where I´m totally out of Ideas and I was hoping that somebody can help me out there. 

I try to explain the problem as best as I can.

I have a global number which refelects a users membership in teams. This number represents the maximum of a given user. A user can then be assigned to teams. The data comes from a database. It is now possible to change the proportion of a team membership a user belongs to, but together they cannot exceed the global limit of the current user. 

I´m currently totally out of ideas. The validation should check if the teamValues together do not exceed the global total number. I tried to use useFieldArray for that example which works pretty well but I have no idea how I should build up the component and the validation.

I provided a code sandbox with a rudimentary example which reflects what I currently have. Maybe someone can help me out here!

Thanks a lot!!

Link to codesandbox:

https://codesandbox.io/s/lingering-mountain-2is19?file=/src/App.js

Best regards
Tom

**Comment by DrBlackBird** [comment]
Does nobody have an idea? I tried a lot of things but I cannot get the validation working :(.

**Reply by thanh-nguyen-95** [reply]
Please provide at least a working version with one approach that you're most confident (or comfortable) with.

The current CodeSandbox is way too vague:

- No submit button (?)
- What does the "Save" button do on each row?
- How user would user fill the form?
- What would be your approach?

All we know right now is `sumOfTeamValues <= globalValue`. Also, use the `mockData` directly as default values if you can to focus on making the validation working right now. You can add data fetching, reseting, ... later on.

**Reply by DrBlackBird** [reply]
HJi @thanh-nguyen-95 Thanks a lot for yor help! Well basically each "save" button needs to be a submit button. You have to imagine that I load up a user on my page and the assignments are lazy loaded afterwards. I need to be able to modify each assignment on its own but also be aware of the the other assignment values so they dont exceed the global total and vice versa. 

I have made a codesandbox yesterday without useFieldArray because I did not manage it to get it working but I´d like to share my solution. Even if its not perfect it does exactly what I want. If you have any suggestions on how to improve it I would totally appreciate it :).

https://codesandbox.io/s/kind-moore-4lgio?file=/src/Components/TeamAssignment/index.js

**Reply by thanh-nguyen-95** [reply]
I see quite a few problems with your implemenation:

- Make sure to pass `name`, `ref`, `value` and `onChange` to your `TextField` for `react-hook-form` to work correctly:

   ```jsx
   <TextField
     name={name}
     inputRef={ref}
     value={value}
     onChange={onChange}
   />
   ```
- Inconsistent naming strategy which leads to incorrect props passed in (`teamAsignmentId` is one of them, which isn't necessary since you passed in `asignment` already)
- Incorrect assignment sum due to incorrect prop name. If you can use TypeScript, it will help you avoid these problems
- The validation is not up-to-date in some cases. For example, if I change the 2nd assignment without saving, then the validation is valid when I input `0.7` and `0.4`, which is incorrect. The reason is because you're validating values based on async data without updating it in the client. You can solve this by only allowing editing 1 field at a time using a toggle and updating the data everytime user hit 'Save'.

Not bad solution overall. My suggestions are:

- Keep all data (including `globalValue` and `assignmentValues`) in one place
- Have a seperate check (not using Yup) for the relation between `globalValue` and  `assignmentValue`
- Mark an assignment value input as invalid if it is `touched` and exceeds the limit (it doesn't matter which one was changed first/last in my opinion)

Those are based on my experience and personal preference. Implement it however you or your team see fit. Good luck

**Reply by bluebill1049** [reply]
thanks @thanh-nguyen-95 for the detailed answer, could you move into a separate comment and let's mark it as answer.

**Reply by DrBlackBird** [reply]
Hi @thanh-nguyen-95 thanks a lot for very detailed explanation! That helps a lot! I have took care of the first two problems in my production code. 
Typescript is currently not an option. But the "global total" in my production code is delivered from the database also, I hope that no type erros will occur. But I will definitely double check that and try to learn typescript for the future. 

I like the idea of your 4th suggestion. I need to think about of the implemention. Is there a possibility with react-hook-form the do validation "on touch". So if someone was editing a field without saving and tries to modify a 2nd field I wanted to display an error message that changes need to be saved first. But afaik I could currently do validation only on submit. That would help a lot :-). 

And thanks that my solution is not completely horrible 👍 as a beginner that means a lot!

**Reply by thanh-nguyen-95** [reply]
Validation event can be changed using `mode` option. See here: https://react-hook-form.com/api/useform/

If you want to check if a field was touched or its value was changed, try `isDirty` or `isTouched` from `fieldState`. Ref: https://react-hook-form.com/api/usecontroller

Edit: fix typos

**Comment by thanh-nguyen-95** [comment, marked as answer]
https://github.com/react-hook-form/react-hook-form/discussions/5447#discussioncomment-865085

Check this reply for suggestions
